### ISSUE TITLE
not flexible how to update existing content (via pick primary key + rename behaviour)

### ISSUE THREAD

**Issue by djay** [collective/plone.importexport#9]
Currently CSV upload requires 4 primary keys. 
- [x] Should **only require 1 key (or none if new content to be uploaded)**. 
- [x] Should also **allow specifying a custom field as the primary key** e.g. if Person object and there is a custom field for "email". 

- [x] User should also get the **option to specify what happens in the case of a match**, e.g. replace, update, ignore, rename old, rename new etc.
- [ ] User query widget to specify only the content you want updated
- [ ] Also allow a **preview to see if anything does conflict or require name changes**.

**Comment by djay** [comment]
proposed UI in https://github.com/collective/plone.importexport/issues/15

**Comment by pigeonflight** [comment]
@djay, @instification reviewing this now.
Have a working build on my dev system.
Changed the readme and I'm met with failing tests (https://github.com/collective/plone.importexport/pull/21)

For the meantime I'll continue to work on my dev system.

**Comment by instification** [comment]
@pigeonflight the tests were already failling - see #19 

Might be worth seeing how bad the failures are as if we can get them working again it would be ideal.

**Comment by pigeonflight** [comment]
So the "setting" that mentions the 4 keys is here... https://github.com/collective/plone.importexport/blob/master/src/plone/importexport/browser/importexport.py#L46

**Comment by pigeonflight** [comment]
> @pigeonflight the tests were already failling - see #19
> 
> Might be worth seeing how bad the failures are as if we can get them working again it would be ideal.

@instification I took at look at the failure. If we're lucky, most of it is due to configuration that expects more than this package provides (like a build of the docs).
I could spend an hour poking at it.

**Comment by pigeonflight** [comment]
Made some changes to the travis.yml, trying another pull request https://github.com/collective/plone.importexport/pull/22

**Comment by pigeonflight** [comment]
@instification the code actually needs a lot of linting... I think it's best to keep working on the functionality and lint as I go. Probably won't be passing before Wednesday.

**Comment by pigeonflight** [comment]
I think the "sensible default behaviour" should be to "update in place" on a match.
This is important to define since we won't be implementing flexible options on matching see: #2 
until a "second pass".

@b4oshany @djay @instification

**Comment by pigeonflight** [comment]
mockup of interface 
![ui](https://user-images.githubusercontent.com/31827/51917146-e4b62480-23ac-11e9-9072-082c9d9ae1c1.png)

**Comment by djay** [comment]
Yes update in place seems best default.

**Comment by pigeonflight** [comment]
Just a quick note. 
I did a test with a csv file.

I can confirm that imports work when there is no UID column in the csv. I previously had the misunderstanding that UID was a needed primary key.

The 'keys' that appear to block imports are `path` and `@type`:
The code that blocks the imports is here: https://github.com/collective/plone.importexport/blob/master/src/plone/importexport/browser/importexport.py#L214-L219

Finally, I haven't tested this, there seems to be code that blocks the import if the parent is missing: https://github.com/collective/plone.importexport/blob/master/src/plone/importexport/browser/importexport.py#L227-L230

**Comment by pigeonflight** [comment]
**Without a path column we would be limited to "flat" imports**
With the current code, the path is needed to name the file or folder. 
In order to get away without using the path we'd need to have a different approach to naming the file.
Without the path there are some implications.

1. There is no way to import trees, only flat imports would be possible when a path is not defined
2. The target context would be assumed to be './', meaning the current container.

There is already a mechism for generating ids when there is no title or id given, however the resulting import is pretty useless.
See screenshot
![image](https://user-images.githubusercontent.com/31827/51986981-6ff7ee80-246f-11e9-84fd-6edaf903e4a8.png)

**Comment by djay** [comment]
Depends what you mean by useless. It might be to update existing data. It
might be to upload a bunch for form results or news items.

**Comment by b4oshany** [comment]
> Finally, I haven't tested this, there seems to be code that blocks the import if the parent is missing: 
Yes, the code blocks the import if the parent is missing. Tested it. I was thinking of creating a default folder for the parent, but I don't think it would be a good implementation knowing that there are other portal types that may have folderish behaviour enabled.

> The 'keys' that appear to block imports are path and @type:
This is also true.
```python
# MUST_INCLUDED_ATTRIBUTES = ['@type', 'path', 'id', 'UID']
MUST_INCLUDED_ATTRIBUTES = ['@type', 'path']
```
